### PR TITLE
Surface refs increment/decrement on context creation/destruction

### PR DESCRIPTION
**Pull request by Rubo3**
I moved the code used by `vkvg_surface_destroy` to decrement the surface references to a new function `_surface_dereference`, so that `vkvg_destroy` could make use of it.
Based on this comment: https://github.com/jpbruyere/vkvg/pull/87#issuecomment-1047919175.

### CODE REVIEW

**Comment by jpbruyere** [comment]
Sorry, but there are too much white space changes, also the ref +/- and test must be guarded by the same mutex. I don't think a separate _dereference function is needed. Also the name dereference is a typical c/c++ error, so I don't like this name in a  function...But thanks to try so hard to improve vkvg...

**Comment by Rubo3** [comment]
Ok, but so what do we do? We still need to increment and decrement the references, as per your comment. And why is it an error? By symmetry, if we have a `*_reference` function which increments the reference, and we do, a `*_dereference` function which does the opposite to me seems reasonable.

**Comment by jpbruyere** [comment]
> And what's the problem with white spaces? I simply indented the code to look good on one column, using tabs.

It adds unrelated changes in the diffs that make reviewing harder, also for that particular struct (surface_t), the indentation looks good on my side, I don't know what's the problem on yours, maybe you may send a picture of this struct rendered with your editor with white space visualisation enabled, so we may discuss further any white space problem.

**Comment by Rubo3** [comment]
Using GitHub's viewer (and editor):

Screenshot_20220226_112753

**Comment by jpbruyere** [comment]
I use tab=4, here it looks like tab=8, aren't you able to set tab to 4 in this editor config?

**Comment by Rubo3** [comment]
I just found that appending `ts=4` to the URL does the trick.

**Comment by jpbruyere** [comment]
this could be set globaly for your profile: https://github.com/settings/appearance

**Comment by jpbruyere** [comment]
Note that I have to thank you for that discussion on white space, so now I've set tabs to 4 on my github, and code looks better online for me too ;-)

**Comment by Rubo3** [comment]
So, do we have to replicate the same surface unreferencing on `vkvg_destroy` and `vkvg_push_group` since a function is not an option?

**Comment by jpbruyere** [comment]
decrementing ref is done with a call to vkvg_surface_destroy, and incrementing with vkvg_surface_reference, those funcs are not part of the context, (no recording is done) so their may be used safely from withing...

**Comment by Rubo3** [comment]
Ok, but here https://github.com/jpbruyere/vkvg/pull/87#issuecomment-1047919175 you said that the unreferencing has to be done even on the context's destruction, so do we have to copy the code from `vkvg_surface_destroy` over `vkvg_destroy`?

**Comment by Rubo3** [comment]
Right, since its only effect would be to decrement the ref count.

**Comment by Rubo3** [comment]
Can you please reopen this request?

**Comment by jpbruyere** [comment]
Wouldn't be easier to start a fresh one?

**Comment by Rubo3** [comment]
Yes, but this one's already here, so it doesn't take much to clean it.

**Comment by Rubo3** [comment]
I think it's done, I don't know if you want to put `vkvg_surface_destroy` on `release_context_ressources` (which has a typo), since we're not actually freeing it.

**Comment by jpbruyere** [comment]
> I think it's done, I don't know if you want to put `vkvg_surface_destroy` on `release_context_ressources` (which has a typo), since we're not actually freeing it.

surface is not really a ressource, that's the target, ressources are the vulkan objects used internally in the context, and other things, that's a good name for me.

**Comment by Rubo3** [comment]
Yes, but it's "resource", not "ressource".

File: src/vkvg_context.c [hunk 1]
```diff
@@ -112,6 +112,7 @@ VkvgContext vkvg_create(VkvgSurface surf)
 	}
 
 	ctx->pSurf = surf;
+	vkvg_surface_reference (surf);
```
Comment (jpbruyere) [inline src/vkvg_context.c:115]: This is not the right place to put it. It must come after surface state test! Also what if the context is reused? Have you noticed that there are 2 pathes for initializing context. An existing context cached in the device could be reused, surface reference as also to be done there..

**Comment by jpbruyere** [comment]
and look also [here](https://github.com/jpbruyere/vkvg/blob/c5632538ab76fe601e90c7a06e24545210acb3f9/src/vkvg_context.c#L298), context may be store in device for fast reuse, surface ref has also to be decremented in that case.

**Comment by Rubo3** [comment]
> look also [here](https://github.com/jpbruyere/vkvg/blob/c5632538ab76fe601e90c7a06e24545210acb3f9/src/vkvg_context.c#L298)

What if we put `vkvg_surface_destroy (ctx->pSurf);` before that line?

File: src/vkvg_context.c (hunk 1 above)
Comment (Rubo3) [reply src/vkvg_context.c]: Right, what about this https://github.com/Rubo3/vkvg/commit/a0c5684b9d56c54d0063a10f312db8ead498db39?

**Comment by jpbruyere** [comment]
the surf reference +, I would have put it maybe in the _init_ctx, common to both context creation pathes.....

**Comment by Rubo3** [comment]
Right. Better now?

**Pull request merged**